### ISSUE TITLE
Using memcpy on tensor data when the tensor is not contiguous

### ISSUE THREAD

**Issue by audiovention** [ggerganov/ggml#584]
I see in some places in the codebase the implicit assumption that a tensor's memory is contiguous, without any checks/asserts.
For example "ggml_metal_set_tensor" (get_tensor as well) has:
`memcpy((void *) ((uint8_t *) id_dst.contents + offs), t->data, ggml_nbytes(t));`

I just want to confirm - this would produce wrong results if the tensor has non-identity strides right?

**Comment by ggerganov** [comment]
Yes that's correct. In some places the necessary checks / asserts are missing.
